From a2be18d0bd7443681740e8cca880dd350870c922 Mon Sep 17 00:00:00 2001
From: Clark DuVall <cduvall@chromium.org>
Date: Wed, 15 Jan 2025 12:58:23 -0800
Subject: [PATCH] [Merge M132] Catch a couple more common GPU errors to avoid sending crash reports

These errors are happening quite a bit and don't give much useful info
in the crash reports.

(cherry picked from commit 3d3e68023fd6e7d89a87ef450b68bf60f1a3d85e)

Bug: 385396982,388312327
Change-Id: I9d8a0a45661f61ee7e7a14e74ba2b2145b766377
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6138584
Reviewed-by: Ayu Ishii <ayui@chromium.org>
Reviewed-by: Mike Wasserman <msw@chromium.org>
Commit-Queue: Clark DuVall <cduvall@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1401572}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6175803
Commit-Queue: Mike Wasserman <msw@chromium.org>
Auto-Submit: Clark DuVall <cduvall@chromium.org>
Cr-Commit-Position: refs/branch-heads/6834@{#3789}
Cr-Branched-From: 47a3549fac11ee8cb7be6606001ede605b302b9f-refs/heads/main@{#1381561}
---

--- a/services/on_device_model/ml/chrome_ml.cc
+++ b/services/on_device_model/ml/chrome_ml.cc
@@ -49,9 +49,10 @@
 enum class GpuErrorReason {
   kOther = 0,
   kDxgiErrorDeviceHung = 1,
-  kDxgiErrorDeviceRemoved = 2,
+  kDeviceRemoved = 2,
   kDeviceCreationFailed = 3,
-  kMaxValue = kDeviceCreationFailed,
+  kOutOfMemory = 4,
+  kMaxValue = kOutOfMemory,
 };
 
 void FatalGpuErrorFn(const char* msg) {
@@ -60,10 +61,14 @@
   GpuErrorReason error_reason = GpuErrorReason::kOther;
   if (msg_str.find("DXGI_ERROR_DEVICE_HUNG") != std::string::npos) {
     error_reason = GpuErrorReason::kDxgiErrorDeviceHung;
-  } else if (msg_str.find("DXGI_ERROR_DEVICE_REMOVED") != std::string::npos) {
-    error_reason = GpuErrorReason::kDxgiErrorDeviceRemoved;
+  } else if (msg_str.find("DXGI_ERROR_DEVICE_REMOVED") != std::string::npos ||
+             msg_str.find("VK_ERROR_DEVICE_LOST") != std::string::npos) {
+    error_reason = GpuErrorReason::kDeviceRemoved;
   } else if (msg_str.find("Failed to create device") != std::string::npos) {
     error_reason = GpuErrorReason::kDeviceCreationFailed;
+  } else if (msg_str.find("VK_ERROR_OUT_OF_DEVICE_MEMORY") !=
+             std::string::npos) {
+    error_reason = GpuErrorReason::kOutOfMemory;
   }
   base::UmaHistogramEnumeration("OnDeviceModel.GpuErrorReason", error_reason);
   if (error_reason == GpuErrorReason::kOther) {
--- a/tools/metrics/histograms/enums.xml
+++ b/tools/metrics/histograms/enums.xml
@@ -12065,8 +12065,9 @@
 <enum name="GpuErrorReason">
   <int value="0" label="Other"/>
   <int value="1" label="DXGI_ERROR_DEVICE_HUNG"/>
-  <int value="2" label="DXGI_ERROR_DEVICE_REMOVED"/>
+  <int value="2" label="DXGI_ERROR_DEVICE_REMOVED or VK_ERROR_DEVICE_LOST"/>
   <int value="3" label="Failed to create device."/>
+  <int value="4" label="VK_ERROR_OUT_OF_DEVICE_MEMORY"/>
 </enum>
 
 <enum name="GpuImageUsageState">
